stan/math/prim/mat/functor/map_rect_concurrent.hpp

```diff
@@ -20,30 +22,43 @@ namespace internal {
  * the environment variable STAN_NUM_THREADS and follows these
  * conventions:
  *
- * - STAN_NUM_THREADS is not defined or is not a number => num_threads=1
+ * - STAN_NUM_THREADS is not defined => num_threads=1
  * - STAN_NUM_THREADS is positive => num_threads is set to the
  *   specified number
  * - STAN_NUM_THREADS is set to -1 => num_threads is the number of
  *   available cores on the machine
- * - STAN_NUM_THREADS < -1 => num_threads is 1
+ * - STAN_NUM_THREADS < -1, STAN_NUM_THREADS = 0 or STAN_NUM_THREADS is
+ *   not numeric => throws an exception
  *
  * Should num_threads exceed the number of jobs, then num_threads will
  * be set equal to the number of jobs.
  *
  * @param num_jobs number of jobs
  * @return number of threads to use
+ * @throws std::runtime_error if the value of STAN_NUM_THREADS env. variable
+ * is invalid
  */
 inline int get_num_threads(int num_jobs) {
   int num_threads = 1;
 #ifdef STAN_THREADS
   const char* env_stan_num_threads = std::getenv("STAN_NUM_THREADS");
   if (env_stan_num_threads != nullptr) {
-    const int env_num_threads = std::atoi(env_stan_num_threads);
-    if (env_num_threads > 0)
-      num_threads = env_num_threads;
-    else if (env_num_threads == -1)
-      num_threads = std::thread::hardware_concurrency();
-    // anything else will use 1 thread.
+    try {
+      const int env_num_threads
+          = boost::lexical_cast<int>(env_stan_num_threads);
+      if (env_num_threads > 0)
+        num_threads = env_num_threads;
+      else if (env_num_threads == -1)
+        num_threads = std::thread::hardware_concurrency();
+      else
+        boost::throw_exception(
```

Comment:
Could we please use one of the standard functions which we have in stan for this? Those live in `stan/math/prim/scal/err`. Ideally you take one which does throw something different than a `domain_error` since these are recoverable in Stan (although we are here fixing math). I think that the `invalid_argument` function may work here, but please have a look yourself.

Reply:
I was unaware of the error reporting functions. I agree that we do not want a recoverable error. It seems that `invalid_argument` is the best match, although some OpenCL code also uses `system_error` which has some appeal but I don't see how I would define an error code, so I've used `invalid_argument`.